File: android/src/main/java/io/kreador/callkit/CallPush.java
package io.kreador.callkit;

import java.util.Map;

/**
 * Parses a call push's data map. Pure Java so it can be unit tested.
 *
 * Accepts the documented keys (type, callId, callerName, handle, hasVideo) and the
 * legacy iOS payload keys of this plugin (ConnectionId, Username).
 */
public final class CallPush {

    public final String type;
    public final String callId;
    public final String callerName;
    public final String handle;
    public final boolean hasVideo;

    private CallPush(String type, String callId, String callerName, String handle, boolean hasVideo) {
        this.type = type;
        this.callId = callId;
        this.callerName = callerName;
        this.handle = handle;
        this.hasVideo = hasVideo;
    }

    /** @return null when the data is not a call push. */
    public static CallPush parse(Map<String, String> data) {
        if (data == null) {
            return null;
        }
        String callId = first(data, "callId", "uuid", "ConnectionId");
        if (callId == null) {
            return null;
        }

        String type = data.get("type");
        if (type == null) {
            type = CallMessagingService.TYPE_INCOMING_CALL;
        }
        if (!CallMessagingService.TYPE_INCOMING_CALL.equals(type) && !CallMessagingService.TYPE_CALL_ENDED.equals(type)) {
            return null;
        }

        String callerName = first(data, "callerName", "name", "Username");
        String handle = first(data, "handle", "number");
        if (callerName == null) {
            callerName = handle != null ? handle : "Unknown";
        }
        if (handle == null) {
            handle = callerName;
        }

        return new CallPush(type, callId, callerName, handle, "true".equalsIgnoreCase(data.get("hasVideo")));
    }

    private static String first(Map<String, String> data, String... keys) {
        for (String key : keys) {
            String value = data.get(key);
            if (value != null && !value.isEmpty()) {
                return value;
            }
        }
        return null;
    }
}

File: ios/Sources/CallKitPlugin/CallPush.swift
import Foundation

/// Parses a call push payload (PushKit `dictionaryPayload`). Mirrors Android's `CallPush`.
///
/// Accepts the documented keys (`type`, `callId`, `callerName`, `handle`, `hasVideo`) and the
/// legacy keys of the original iOS plugin (`ConnectionId`, `Username`).
public struct CallPush: Equatable {
    public static let typeIncomingCall = "incoming_call"
    public static let typeCallEnded = "call_ended"

    public let type: String
    public let callId: String
    public let callerName: String
    public let handle: String
    public let hasVideo: Bool

    /// - Returns: `nil` when the payload is not a call push.
    public static func parse(_ payload: [AnyHashable: Any]) -> CallPush? {
        guard let callId = first(payload, ["callId", "uuid", "ConnectionId"]) else {
            return nil
        }
        let type = string(payload["type"]) ?? typeIncomingCall
        guard type == typeIncomingCall || type == typeCallEnded else {
            return nil
        }
        let rawHandle = first(payload, ["handle", "number"])
        let callerName = first(payload, ["callerName", "name", "Username"]) ?? rawHandle ?? "Unknown"
        let handle = rawHandle ?? callerName
        return CallPush(type: type, callId: callId, callerName: callerName, handle: handle, hasVideo: bool(payload["hasVideo"]))
    }

    static func string(_ value: Any?) -> String? {
        if let text = value as? String {
            return text.isEmpty ? nil : text
        }
        if let number = value as? NSNumber {
            return number.stringValue
        }
        return nil
    }

    static func bool(_ value: Any?) -> Bool {
        if let flag = value as? Bool {
            return flag
        }
        if let text = value as? String {
            return text.lowercased() == "true" || text == "1"
        }
        if let number = value as? NSNumber {
            return number.boolValue
        }
        return false
    }

    private static func first(_ payload: [AnyHashable: Any], _ keys: [String]) -> String? {
        for key in keys {
            if let value = string(payload[key]) {
                return value
            }
        }
        return nil
    }
}
